smart-contract-feature-development.md

```diff
@@ -0,0 +1,94 @@
+then the original contract may need to have the `virtual` keyword added in places. The `virtual` keyword should only be added in
+places where it must be added and nowhere in addition.
+
+### Handling Semver
+
+The parent contract should update its `version` getter to be a real function. The child contract should implement a `version` getter like so:
+
+```solidity
+function version() public override returns (string) {
+  return string.concat(super.version(), "+feature-name");
+}
+```
+
+The `feature-name` should be some identifier that uniquely identifies the feature. It could also be a hardfork name if the release is tied to a hard fork.
+
+### Handling Storage
+
+Handling storage becomes complex and there is no solution that handles all cases well. We generally want to ensure that the contracts
+are forwards compatible, so if `solc` is managing storage, the base contract should have spacers when we know that a feature contract
+is scheduled to be merged soon.
+
+Another solution is to use unstructured storage, meaning not letting `solc` manage storage and instead using custom storage slots.
+Assuming that digests and unique preimages are used for the storage slot keys, then this can work well for ensuring there are no
+storage slot collisions. This works well if there is no need for `solc` to manage the storage layout.
+
+If there are multiple feature contracts based on the same base contract being worked on in parallel, then there is complexity around
+layout management and it is up to the teams working on features to coordinate with each other if they risk mangling storage layouts.
+Using unstructured storage greatly reduces the coordination overhead here.
+
+More attention should be paid when the contracts are going to be deployed to a production network. It is less important to have the
+storage layouts 100% perfect when the contracts are going to be used on a transient network like a devnet. It is up to the developers
+to be mindful of their own upgrade paths as projects may have varying requirements.
+
+Given enough time, there will be storage layout issues. In the worst case, they can be fixed at upgrade time using the `StorageSetter`
+contract, this pattern is already used in production for `Initializable`.
+
+### Deploy Script
+
+The deploy script should be able to parse the deploy config in a standard way where the same config option enables the feature across the
+entire codebase. This means that a single config key/value pair should be able to have predeploys set in the L2 genesis, have the feature
+enabled contracts deployed to L1 and also tell op-node/op-geth to use that particular feature.
+
+### Merging Contracts
+
+The feature specific contract should be merged back into the base contract when the scheduling allows. This means that the code is nearing
+ready to be deployed to production and there is nothing immediately before that is going to be deployed and released.
```

Comment:
I think we consistently see that requiring strict ordering of changes becomes really problematic because there are so often unexpected things.  Fault Proofs for example was "done" when an RC was taken but then further changes were required.  Other changes had already been merged in though so we had to cherry-pick commits which introduces risks of missing some important commits.

For me, the really key property we need is to keep develop releasable at all times (our tests and release process tries to disprove the theory that it's ok to release by finding issues). I think this inheritance approach adds a useful tool to our toolbox so we can get closer to that goal, but it seems like it doesn't entirely solve the problem if we are still needing to manage this sequencing. That said, if we only do this merge once a feature is confirmed ready to ship then it's fine because anything else that needs to ship is safe to deploy the merged contracts.